codec/src/test/java/io/netty/handler/codec/ByteToMessageDecoderTest.java

```diff
@@ -345,4 +344,176 @@ public CompositeByteBuf addComponent(boolean increaseWriterIndex, ByteBuf buffer
+            this.messageSize = messageSize;
+            setCumulator(new PartialCumulator());
+        }
+
+        class PartialCumulator implements Cumulator {
+            @Override
+            public ByteBuf cumulate(ByteBufAllocator alloc, ByteBuf cumulation, ByteBuf in) {
+                ++cumulations;
+                if (!cumulation.isReadable()) {
+                    cumulation.release();
+                    return in;
+                }
+
+                assertTrue(cumulation.readableBytes() < messageSize);
+                ByteBuf result = alloc.buffer(messageSize);
+                result.writeBytes(cumulation);
+                while (result.isWritable() && in.isReadable()) {
+                    result.writeByte(in.readByte());
+                }
+
+                cumulation.release();
+                if (!in.isReadable()) {
+                    in.release();
+                }
+
+                return result;
+            }
+        }
+    }
+
+    @Test
+    public void partialCumulate() {
+        final Object decoded = new Object();
+        for (int firstMessageCount = 0 ; firstMessageCount <= 1 ; ++firstMessageCount) {
+            for (int unreadFinalBytes = 0 ; unreadFinalBytes <= 1 ; ++unreadFinalBytes) {
+                final PartialCumulationDecoder decoder = new PartialCumulationDecoder(2) {
+                    @Override
+                    protected void decode(ChannelHandlerContext ctx, ByteBuf in, List<Object> out) {
+                        while (in.readableBytes() >= 2) {
+                            in.skipBytes(2);
+                            out.add(decoded);
+                        }
+                    }
+                };
+
+                ByteBuf in1 = Unpooled.buffer().writeZero(firstMessageCount * 2 + 1);
+                ByteBuf in2 = Unpooled.buffer().writeZero(5 + unreadFinalBytes);
+                EmbeddedChannel channel = new EmbeddedChannel(decoder);
+
+                assertEquals(firstMessageCount > 0, channel.writeInbound(in1));
+                for (int i = 0 ; i < firstMessageCount ; ++i) {
+                    assertSame(decoded, channel.readInbound());
+                }
+                assertNull(channel.readInbound());
+                assertTrue(channel.writeInbound(in2));
+                for (int i = 0 ; i < 3 ; ++i) {
+                    assertSame(decoded, channel.readInbound());
+                }
+                assertEquals(0, in1.refCnt());
+                assertEquals(2 + unreadFinalBytes, decoder.cumulations);
+                if (unreadFinalBytes > 0) {
+                    assertSame(in2, decoder.cumulation);
+                    assertEquals(1, decoder.cumulation.readableBytes());
+                    assertFalse(channel.finish());
+                } else {
+                    assertEquals(0, in2.refCnt());
+                    assertNull(decoder.cumulation);
+                    assertFalse(channel.finish());
```

Comment:
nit: you can remove this out of the else block and remove it from the if block.

Reply:
Yep, of course